# Capstone Notebook
This notebook will be used to complete the final course of the IBM professional data science certification program

In [1]:
import pandas as pd
import numpy as np
import urllib.request
import bs4 as bs

### Scrape web page and decode contents

In [6]:

url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
req = urllib.request.urlopen(url)
article = req.read().decode()

### Filter contents, save table contents to DataFrame

In [82]:
#source = urllib.request.urlopen('http://www.gcoins.net/en/catalog/view/45518').read()
soup = bs.BeautifulSoup(article,'lxml')

table = soup.find('table')
table_rows = table.find_all('tr')

res = []
for tr in table_rows:
    td = tr.find_all('td')
    row = [tr.text.strip() for tr in td if tr.text.strip()]
    if row:
        res.append(row)


df = pd.DataFrame(res, columns=["Postcode", "Borough", "Neighbourhood"])
df.head()


Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [83]:
df.shape

(289, 3)

### Remove Not Assigned Boroughs

In [85]:
df = df[df.Borough != 'Not assigned']
df.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights

In [86]:
df.shape

(212, 3)

**77 Rows reduced [289 - 212]**

### Group rows by Postcode and Concatenate Neighbourhood

In [103]:
df2 = pd.DataFrame({'Neighbourhood' : df.groupby(['Postcode', 'Borough'])['Neighbourhood'].agg(lambda col: ', '.join(col))}).reset_index()
df2.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

### Copy Borough value to Neighbourhood for Not Assigned Neighbourhoods

In [88]:
df2.loc[df2['Neighbourhood'] == 'Not assigned', 'Neighbourhood'] = df2.loc[df2['Neighbourhood'] == 'Not assigned'].Borough
df2[df2['Postcode'] == 'M7A']

Postcode       Borough Neighbourhood
85      M7A  Queen's Park  Queen's Park

In [89]:
df2.shape

(103, 3)

### Import Coordinates and join on existing DF
**Coordinate data downloaded from <a href="https://cocl.us/Geospatial_data">https://cocl.us/Geospatial_data</a>**

In [90]:
coords = pd.read_csv('geo/Geospatial_Coordinates.csv')
coords.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [105]:
df3 = df2.join(coords.set_index('Postal Code'), on='Postcode')
df3.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [115]:
df3.shape

(103, 5)

## Cluster and Visualize Data

Import relevant libraries

In [125]:
import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

import requests # library to handle requests

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Find Toronto Geocode

In [108]:
address = 'Toronto'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

/Users/sm/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  This is separate from the ipykernel package so we can avoid doing imports until


The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Add data points to Folium Map

In [113]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

neighbourhoods = df3
# add markers to map
for lat, lng, borough, neighbourhood in zip(neighbourhoods['Latitude'], neighbourhoods['Longitude'], neighbourhoods['Borough'], neighbourhoods['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Filter data to only work with Scarborough

In [188]:
scarborough_data = neighbourhoods[neighbourhoods['Borough'] == 'Scarborough'].reset_index(drop=True)
scarborough_data.head()

Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [118]:
address = 'Scarborough, ON'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Scarboroughare {}, {}.'.format(latitude, longitude))

/Users/sm/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  This is separate from the ipykernel package so we can avoid doing imports until


The geograpical coordinate of Scarboroughare 43.773077, -79.257774.


Plot Scarborough Data, increase zoom

In [120]:
map_scarborough = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(scarborough_data['Latitude'], scarborough_data['Longitude'], scarborough_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_scarborough)  
    
map_scarborough

In [121]:
#CLIENT_ID = ##hidden
#CLIENT_SECRET = ##hidden
VERSION = '20180605' # Foursquare API version

#### Function to Explore Venues 

In [156]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [157]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

scarborough_venues = getNearbyVenues(names=scarborough_data['Neighbourhood'],
                                   latitudes=scarborough_data['Latitude'],
                                   longitudes=scarborough_data['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West, Steeles West
Upper Rouge


#### Perform One hot encoding to allow for clustering of data

In [171]:
# one hot encoding
scarborough_onehot = pd.get_dummies(scarborough_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
scarborough_onehot['Neighbourhood'] = scarborough_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [scarborough_onehot.columns[-1]] + list(scarborough_onehot.columns[:-1])
scarborough_onehot = scarborough_onehot[fixed_columns]
#scarborough_onehot.head()
scarborough_onehot

Neighbourhood  American Restaurant  \
0                                      Rouge, Malvern                    0   
1                                      Rouge, Malvern                    0   
2              Highland Creek, Rouge Hill, Port Union                    0   
3              Highland Creek, Rouge Hill, Port Union                    0   
4                   Guildwood, Morningside, West Hill                    0   
5                   Guildwood, Morningside, West Hill                    0   
6                   Guildwood, Morningside, West Hill                    0   
7                   Guildwood, Morningside, West Hill                    0   
8                   Guildwood, Morningside, West Hill                    0   
9                   Guildwood, Morningside, West Hill                    0   
10                                             Woburn                    0   
11                                             Woburn                    0   
12                                             Woburn                    0   
13                                             Woburn                    0   
14                                          Cedarbrae                    0   
15                                          Cedarbrae                    0   
16                                          Cedarbrae                    0   
17                                          Cedarbrae                    0   
18                                          Cedarbrae                    0   
19                                          Cedarbrae                    0   
20                                          Cedarbrae                    0   
21                                Scarborough Village                    0   
22                                Scarborough Village                    0   
23        East Birchmount Park, Ionview, Kennedy Park                    0   
24        East Birchmount Park, Ionview, Kennedy Park                    0   
25        East Birchmount Park, Ionview, Kennedy Park                    0   
26        East Birchmount Park, Ionview, Kennedy Park                    0   
27        East Birchmount Park, Ionview, Kennedy Park                    0   
28        East Birchmount Park, Ionview, Kennedy Park                    0   
29        East Birchmount Park, Ionview, Kennedy Park                    0   
..                                                ...                  ...   
53                                  Maryvale, Wexford                    0   
54                                  Maryvale, Wexford                    0   
55                                  Maryvale, Wexford                    0   
56                                          Agincourt                    0   
57                                          Agincourt                    0   
58                                          Agincourt                    0   
59                                          Agincourt                    0   
60            Clarks Corners, Sullivan, Tam O'Shanter                    0   
61            Clarks Corners, Sullivan, Tam O'Shanter                    0   
62            Clarks Corners, Sullivan, Tam O'Shanter                    0   
63            Clarks Corners, Sullivan, Tam O'Shanter                    0   
64            Clarks Corners, Sullivan, Tam O'Shanter                    0   
65            Clarks Corners, Sullivan, Tam O'Shanter                    0   
66            Clarks Corners, Sullivan, Tam O'Shanter                    0   
67            Clarks Corners, Sullivan, Tam O'Shanter                    0   
68            Clarks Corners, Sullivan, Tam O'Shanter                    0   
69  Agincourt North, L'Amoreaux East, Milliken, St...                    0   
70  Agincourt North, L'Amoreaux East, Milliken, St...                    0   
71                      L'Amoreaux West, Steeles West                    0   
72                      L'Amoreaux West, Steeles West         

In [172]:
scarborough_grouped = scarborough_onehot.groupby('Neighbourhood').mean().reset_index()
scarborough_grouped.shape

(16, 54)

#### Display top venues for each neighbourhood

In [137]:
num_top_venues = 5

for hood in scarborough_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = scarborough_grouped[scarborough_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                 venue  freq
0               Lounge  0.25
1       Breakfast Spot  0.25
2         Skating Rink  0.25
3       Clothing Store  0.25
4  American Restaurant  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                       venue  freq
0                       Park   0.5
1                 Playground   0.5
2        American Restaurant   0.0
3          Korean Restaurant   0.0
4  Latin American Restaurant   0.0


----Birch Cliff, Cliffside West----
                   venue  freq
0        College Stadium  0.25
1           Skating Rink  0.25
2  General Entertainment  0.25
3                   Café  0.25
4              Pet Store  0.00


----Cedarbrae----
                venue  freq
0              Bakery  0.14
1                Bank  0.14
2     Thai Restaurant  0.14
3  Athletics & Sports  0.14
4    Hakka Restaurant  0.14


----Clairlea, Golden Mile, Oakridge----
          venue  freq
0        Bakery  0.22
1      Bus Line  0.22
2  Intersec

In [138]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [141]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = scarborough_grouped['Neighborhood']

for ind in np.arange(scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood  1st Most Common Venue  \
0                                           Agincourt           Skating Rink   
1   Agincourt North, L'Amoreaux East, Milliken, St...                   Park   
2                         Birch Cliff, Cliffside West  General Entertainment   
3                                           Cedarbrae        Thai Restaurant   
4                     Clairlea, Golden Mile, Oakridge                 Bakery   
5             Clarks Corners, Sullivan, Tam O'Shanter            Pizza Place   
6     Cliffcrest, Cliffside, Scarborough Village West    American Restaurant   
7   Dorset Park, Scarborough Town Centre, Wexford ...      Indian Restaurant   
8         East Birchmount Park, Ionview, Kennedy Park         Discount Store   
9                   Guildwood, Morningside, West Hill      Electronics Store   
10             Highland Creek, Rouge Hill, Port Union                    Bar   
11                      L'Amoreaux West, Steeles West   Fast Food Restaurant   
12                                  Maryvale, Wexford                 Bakery   
13                                     Rouge, Malvern   Fast Food Restaurant   
14                                Scarborough Village             Playground   
15                                             Woburn            Coffee Shop   

   2nd Most Common Venue      3rd Most Common Venue  \
0         Breakfast Spot                     Lounge   
1             Playground         Chinese Restaurant   
2           Skating Rink                       Café   
3     Athletics & Sports                     Bakery   
4               Bus Line               Intersection   
5           Noodle House         Chinese Restaurant   
6                  Motel                Coffee Shop   
7              Pet Store  Latin American Restaurant   
8          Train Station                Coffee Shop   
9         Breakfast Spot        Rental Car Location   
10         Moving Target      Vietnamese Restaurant   
11    Chinese Restaurant              Grocery Store   
12            Smoke Shop             Breakfast Spot   
13            Print Shop      Vietnamese Restaurant   
14      Business Service      Vietnamese Restaurant   
15     Korean Restaurant          Indian Restaurant   

        4th Most Common Venue  5th Most Common Venue  6th Most Common Venue  \
0              Clothing Store  Vietnamese Restaurant            Coffee Shop   
1               Grocery Store  General Entertainment    Fried Chicken Joint   
2             College Stadium  Vietnamese Restaurant            Coffee Shop   
3                        Bank       Hakka Restaurant    Fried Chicken Joint   
4                Soccer Field   Fast Food Restaurant          Metro Station   
5             Thai Restaurant    Fried Chicken Joint     Italian Restaurant   
6            Hakka Restaurant          Grocery Store  General Entertainment   
7          Chinese Restaurant  Vietnamese Restaurant                    Bar   
8                  Hobby Shop            Bus Station       Department Store   
9                 Pizza Place         Medical Center     Mexican Restaurant   
10                Coffee Shop       Hakka Restaurant          Grocery Store   
11                Coffee Shop    Japanese Restaurant         Sandwich Place   
12  Middle Eastern Restaurant  Vietnamese Restaurant            Coffee Shop   
13             Clothing Store       Hakka Restaurant          Grocery Store   
14                Coffee Shop       Hakka Restaurant          Grocery Store   
15      Vietnamese Restaurant       Hakka Restaurant          Grocery Store   

    7th Most Common Venue  8th Most Common Venue  9th Most Common Venue  \
0           Grocery Store  General Entertainment    Fried Chicken Joint   
1    Fast Food Restaurant      Electronics Store         Discount Store   
2        Hakka Restaurant          Grocery Store    Fried Chicken Joint   
3    Caribbean Restaurant  Vietnamese Restaurant        College Stadium   
4                    Park

### Perform Clustering Based on Above Data

In [179]:
# set number of clusters
kclusters = 5

scarborough_grouped_clustering = scarborough_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(scarborough_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 3, 1, 1, 1, 1, 2, 1, 1, 1], dtype=int32)

In [201]:

scarborough_merged = scarborough_data 

# add clustering labels
scarborough_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
scarborough_merged = scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

scarborough_merged # check the last columns!

Postcode      Borough                                      Neighbourhood  \
0       M1B  Scarborough                                     Rouge, Malvern   
1       M1C  Scarborough             Highland Creek, Rouge Hill, Port Union   
2       M1E  Scarborough                  Guildwood, Morningside, West Hill   
3       M1G  Scarborough                                             Woburn   
4       M1H  Scarborough                                          Cedarbrae   
5       M1J  Scarborough                                Scarborough Village   
6       M1K  Scarborough        East Birchmount Park, Ionview, Kennedy Park   
7       M1L  Scarborough                    Clairlea, Golden Mile, Oakridge   
8       M1M  Scarborough    Cliffcrest, Cliffside, Scarborough Village West   
9       M1N  Scarborough                        Birch Cliff, Cliffside West   
10      M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...   
11      M1R  Scarborough                                  Maryvale, Wexford   
12      M1S  Scarborough                                          Agincourt   
13      M1T  Scarborough            Clarks Corners, Sullivan, Tam O'Shanter   
14      M1V  Scarborough  Agincourt North, L'Amoreaux East, Milliken, St...   
15      M1W  Scarborough                      L'Amoreaux West, Steeles West   

     Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0   43.806686 -79.194353               1   Fast Food Restaurant   
1   43.784535 -79.160497               3                    Bar   
2   43.763573 -79.188711               1      Electronics Store   
3   43.770992 -79.216917               1            Coffee Shop   
4   43.773136 -79.239476               1        Thai Restaurant   
5   43.744734 -79.239476               1             Playground   
6   43.727929 -79.262029               2         Discount Store   
7   43.711112 -79.284577               1                 Bakery   
8   43.716316 -79.239476               1    American Restaurant   
9   43.692657 -79.264848               1  General Entertainment   
10  43.757410 -79.273304               4      Indian Restaurant   
11  43.750072 -79.295849               1                 Bakery   
12  43.794200 -79.262029               1           Skating Rink   
13  43.781638 -79.304302               0            Pizza Place   
14  43.815252 -79.284577               3                   Park   
15  43.799525 -79.318389               1   Fast Food Restaurant   

   2nd Most Common Venue      3rd Most Common Venue  \
0             Print Shop      Vietnamese Restaurant   
1          Moving Target      Vietnamese Restaurant   
2         Breakfast Spot        Rental Car Location   
3      Korean Restaurant          Indian Restaurant   
4     Athletics & Sports                     Bakery   
5       Business Service      Vietnamese Restaurant   
6          Train Station                Coffee Shop   
7               Bus Line               Intersection   
8                  Motel                Coffee Shop   
9           Skating Rink                       Café   
10             Pet Store  Latin American Restaurant   
11            Smoke Shop             Breakfast Spot   
12        Breakfast Spot                     Lounge   
13          Noodle House         Chinese Restaurant   
14            Playground         Chinese Restaurant   
15    Chinese Restaurant              Grocery Store   

        4th Most Common Venue  5th Most Common Venue  6th Most Common Venue  \
0              Clothing Store       Hakka Restaurant          Grocery Store   
1                 Coffee Shop       Hakka Restaurant          Grocery Store   
2                 Pizza Place         Medical Center     Mexican Restaurant   
3       Vietnamese Restaurant       Hakka Restaurant          Grocery Store   
4                        Bank       Hakka Restaurant    Fried Chicken Joint   
5                 Coffee Shop       Hakka Restaurant          Grocery Store   
6                  Hobby Shop 

### Map Clusters based on above

In [200]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(scarborough_merged['Latitude'], scarborough_merged['Longitude'], scarborough_merged['Neighbourhood'], scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Discussion
 There are five clusters. By looking at the data and corresponding clusters, neighbourhoods are clustered into the following categories:
    0. General food places, restaurants etc.
    1. Variety between shops and restaurants
    2. Majority shops and stores
    3. Parks and entertainment
    4. Regional cuising